docs/pypi/download.md

```diff
@@ -0,0 +1,294 @@
+`rules_python` parses the `whl` file [`METADATA`][metadata].
+Packages can express dependencies via `Requires-Dist` and they can add conditions using
+"environment markers", which represent the Python version, OS, etc.
+
+While the PyPI integration provides reasonable defaults to support most
+platforms and environment markers, the values it uses can be customized in case
+more esoteric configurations are needed.
+
+To customize the values used, you need to do two things:
+1. Define a target that returns {obj}`EnvMarkerInfo`
+2. Set the {obj}`//python/config_settings:pip_env_marker_config` flag to
+   the target defined in (1).
+
+The keys and values should be compatible with the [PyPA dependency specifiers
+specification](https://packaging.python.org/en/latest/specifications/dependency-specifiers/).
+This is not strictly enforced, however, so you can return a subset of keys or
+additional keys, which become available during dependency evaluation.
+
+[metadata]: https://packaging.python.org/en/latest/specifications/core-metadata/
+
+(bazel-downloader)=
+### Bazel downloader and multi-platform wheel hub repository.
+
+```{warning}
+This is currently still experimental and whilst it has been proven to work in quite a few
+environments, the APIs are still being finalized and there may be changes to the APIs for this
+feature without much notice.
+
+The issues that you can subscribe to for updates are:
+* {gh-issue}`260`
+* {gh-issue}`1357`
+```
+
+The {obj}`pip` extension supports pulling information from `PyPI` (or a compatible mirror) and it
+will ensure that the [bazel downloader][bazel_downloader] is used for downloading the wheels.
+
+This provides the following benefits:
+* Integration with the [credential_helper](#credential-helper) to authenticate with private
+  mirrors.
+* Cache the downloaded wheels speeding up the consecutive re-initialization of the repositories.
+* Reuse the same instance of the wheel for multiple target platforms.
+* Allow using transitions and targeting free-threaded and musl platforms more easily.
+* Avoids `pip` for wheel fetching and results in much faster dependency fetching.
+
+To enable the feature specify {attr}`pip.parse.experimental_index_url` as shown in
+the {gh-path}`examples/bzlmod/MODULE.bazel` example.
+
+Similar to [uv](https://docs.astral.sh/uv/configuration/indexes/), one can override the
+index that is used for a single package. By default we first search in the index specified by
+{attr}`pip.parse.experimental_index_url`, then we iterate through the
+{attr}`pip.parse.experimental_extra_index_urls` unless there are overrides specified via
+{attr}`pip.parse.experimental_index_url_overrides`.
+
+When using this feature during the `pip` extension evaluation you will see the accessed indexes similar to below:
+```console
+Loading: 0 packages loaded
+    currently loading: docs/
+    Fetching module extension pip in @@//python/extensions:pip.bzl; starting
+    Fetching https://pypi.org/simple/twine/
```

Comment:
Is there a report_progress() call missing that would help clarify what this `Fetching <url>` is actually doing?

Reply:
I tried adding one and updated docs, let's see.